doc/man/man4/tpiperf.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/tpiperf.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH TPIPERF 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B tpiperf
\- Transport Provider Interface performance module (TPIPERF)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/stropt.h>
.br
.B #include <sys/tpiperf.h>
.HP 8
.BI "" tli_stream " = t_open(" tli_device ", " 0 );
.HP 8
.BI ioctl( tli_stream ", " \fRI_PUSH\fP ", " \*(lqtpiperf\*(rq );
.PD 0
.HP 8
.BI ioctl( tli_stream ", " \fRI_POP\fP ", " \*(lqtpiperf\*(rq );
.BI t_close( tli_stream );
.PD
.\"
.\"
.SH DESCRIPTION
.PP
The
.B tpiperf
module is a STREAMS module that provides a transport user supporting the
.I \(lqTransport Provider Interface\(rq
(TPI)
.[
tpi revision 2.0.0 draft 2
.]
with a high performance read/write interface.
.\"
.\"
.SH USAGE
.PP
.B tpiperf
is only suitable for connection-oriented transport service providers of type
.I T_COTS
or
.IR T_COTS_ORD .
(But, from the discussion below under
.RI \(rq IOCTLS \(rq,
it could easily be extended to work for connectionless transport service
providers.)
.PP
The
.B tpiperf
module is typically pushed on an XTI/TLI stream after it has been opened,
bound and placed in the connected state using
.BR xti (3)
library calls or
.BR tpi (7)
primitives.
Once the stream is in the connected state, the
.B tpiperf
module is pushed.
.PP
Once pushed, the module acts in the default mode, which is to do nothing but
pass messages up and down unaffected.
There are three other modes into which the module can be placed with the use of
input-output control commands to acheive several performance testing
objectives.  These modes are discard mode (discard all received data
messages), echo mode (echo all received data messages back to the sender), and
two generation modes (duplicate sent messages as fast a possible to the
peer).  These modes are described in greater detail under

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125

src/modules/tpiperf.c
/*****************************************************************************

 @(#) File: src/modules/tpiperf.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/tpiperf.c (" PACKAGE_ENVR ") " PACKAGE_DATE;


/*
 *  This is a TPI performance testing  module for SCTP that provides some specialized intput-output
 *  controls meant largely for performance testing purpose only.
 */

#include <sys/os7/compat.h>

/*
 *  These are for TPI definitions.
 */
#if defined HAVE_TIHDR_H
#   include <tihdr.h>
#else
#   include <sys/tihdr.h>
#endif

#define TPIPERF_DESCRIP		"TPI Performance (TPIPERF) STREAMS Module"
#define TPIPERF_EXTRA		"Part of UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define TPIPERF_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define TPIPERF_REVISION	"OpenSS7 src/modules/tpiperf.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define TPIPERF_DEVICE		"SVR 4.2 MP STREAMS TPI Performance Module (TPIPERF)"
#define TPIPERF_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define TPIPERF_LICENSE		"GPL"
#define TPIPERF_BANNER		TPIPERF_DESCRIP		"\n" \
				TPIPERF_EXTRA		"\n" \
				TPIPERF_COPYRIGHT	"\n" \
				TPIPERF_REVISION	"\n" \
				TPIPERF_DEVICE		"\n" \
				TPIPERF_CONTACT
#define TPIPERF_SPLASH		TPIPERF_DEVICE		" - " \
				TPIPERF_REVISION

#ifdef LINUX
MODULE_AUTHOR(TPIPERF_CONTACT);
MODULE_DESCRIPTION(TPIPERF_DESCRIP);
MODULE_SUPPORTED_DEVICE(TPIPERF_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(TPIPERF_LICENSE);
#endif				/* MODULE_LICENSE */
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-tpiperf");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#ifndef TPIPERF_MOD_NAME
#define TPIPERF_MOD_NAME	"tpiperf"
#endif				/* TPIPERF_MOD_NAME */

#ifndef TPIPERF_MOD_ID
#define TPIPERF_MOD_ID		0
#endif				/* TPIPERF_MOD_ID */

/*
 *  =========================================================================
 *
 *  STREAMS Definitions
 *
 *  =========================================================================
 */

#define MOD_ID		TPIPERF_MOD_ID
#define MOD_NAME	TPIPERF_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	TPIPERF_BANNER
#else				/* MODULE */
#define MOD_BANNER	TPIPERF_SPLASH
#endif				/* MODULE */

static struct module_info tpiperf_minfo = {
	.mi_idnum = MOD_ID,		/* Module ID number */
	.mi_idname = MOD_NAME,		/* Module name */
	.mi_minpsz = 0,			/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,		/* Max packet size accepted */
	.mi_hiwat = 1,			/* Hi water mark */
	.mi_lowat = 0,			/* Lo water mark */
};

static struct module_stat tpiperf_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat tpiperf_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static streamscall int tpiperf_open(queue_t *, dev_t *, int, int, cred_t *);
static streamscall int tpiperf_close(queue_t *, int, cred_t *);

static streamscall int tpiperf_rput(queue_t *, mblk_t *);
static streamscall int tpiperf_rsrv(queue_t *);

static struct qinit tpiperf_rinit = {
	.qi_putp = tpiperf_rput,	/* Read put (message from below) */
	.qi_srvp = tpiperf_rsrv,
	.qi_qopen = tpiperf_open,	/* Each open */
	.qi_qclose = tpiperf_close,	/* Last close */
	.qi_minfo = &tpiperf_minfo,	/* Information */
	.qi_mstat = &tpiperf_rstat,	/* Statistics */
};

static streamscall int tpiperf_wput(queue_t *, mblk_t *);
static streamscall int tpiperf_wsrv(queue_t *);

static struct qinit tpiperf_winit = {
	.qi_putp = tpiperf_wput,	/* Write put (message from above) */
	.qi_srvp = tpiperf_wsrv,
	.qi_minfo = &tpiperf_minfo,	/* Information */
	.qi_mstat = &tpiperf_wstat,	/* Statistics */
};

static struct streamtab tpiperfinfo = {
	.st_rdinit = &tpiperf_rinit,	/* Upper read queue */
	.st_wrinit = &tpiperf_winit,	/* Upper write queue */
};

/*
 *  =========================================================================
 *
 *  Private Datastructure ctors and dtors
 *
 *  =========================================================================
 */
struct tpiperf {
	queue_t *rq;			/* module read queue */
	queue_t *wq;			/* module write queue */
	mblk_t *mp;			/* stored message */
	uint mode;			/* operations mode */
	uint number;			/* number of messages to send */
	uint burst;			/* number of messages to burst */
	uint streams;			/* number of streams on which to send */
	uint stream;			/* current stream number */
	uint count;			/* current message count */
	uint bytes;			/* current byte count */
	bcid_t bcid;			/* bufcall */
};

static kmem_cachep_t tpiperf_priv_cachep = NULL;

static int
tpiperf_init_caches(void)
{
	if (!tpiperf_priv_cachep
	    && !(tpiperf_priv_cachep =
		 kmem_create_cache(MOD_NAME, sizeof(struct tpiperf), 0, SLAB_HWCACHE_ALIGN, NULL, NULL)
		 )) {
		cmn_err(CE_WARN, "%s: %s: Cannot allocate tpiperf_priv_cachep", MOD_NAME,
			__FUNCTION__);
		return (-ENOMEM);
	}
	return (0);
}

static int
tpiperf_term_caches(void)
{
	if (tpiperf_priv_cachep) {
#ifdef HAVE_KTYPE_KMEM_CACHE_T_P
		if (kmem_cache_destroy(tpiperf_priv_cachep)) {
			cmn_err(CE_WARN, "%s: %s: did not destroy tpiperf_priv_cachep", MOD_NAME,
				__FUNCTION__);
			return (-EBUSY);
		}
#else
		kmem_cache_destroy(tpiperf_priv_cachep);
#endif
	}
	return (0);
}

static struct tpiperf *
tpiperf_alloc_priv(queue_t *q)
{
	struct tpiperf *priv;

	if ((priv = kmem_cache_alloc(tpiperf_priv_cachep, GFP_ATOMIC))) {
		priv->rq = q;
		priv->wq = WR(q);
		priv->mode = 0;
		priv->streams = 1;
		priv->number = -1;
		priv->stream = 0;
		priv->count = 0;
		priv->bytes = 0;
		priv->mp = NULL;
		priv->bcid = 0;
		priv->burst = 0;
		/* Our minpsz, maxpsz matter because we are the first queue under the stream head.
		   We do not want to alter the characteristics of the transport packet sizes so we
		   copy them here. This will allow the stream head to exhibit the same behaviour as
		   before we were pushed. */
		priv->wq->q_minpsz = priv->wq->q_next->q_minpsz;
		priv->wq->q_maxpsz = priv->wq->q_next->q_maxpsz;
		q->q_ptr = WR(q)->q_ptr = priv;
	}
	return (priv);
}

static void
tpiperf_free_priv(queue_t *q)
{
	struct tpiperf *priv;

	if ((priv = (typeof(priv)) q->q_ptr)) {
		if (priv->bcid)
			unbufcall(xchg(&priv->bcid, 0));
		if (priv->mp)
			freemsg(xchg(&priv->mp, NULL));
		q->q_ptr = WR(q)->q_ptr = NULL;
		kmem_cache_free(tpiperf_priv_cachep, priv);
	}
	return;
}

static streamscall void
tpiperf_bufcall(long arg)
{
	queue_t *q = (queue_t *) arg;

	qenable(q);
	return;
}

static streamscall int
tpiperf_wput(queue_t *q, mblk_t *mp)
{
	struct tpiperf *priv = (typeof(priv)) q->q_ptr;
	int db_type = mp->b_datap->db_type;

	if (unlikely((q->q_first || (q->q_flag & QSVCBUSY)) && db_type < QPCTL)) {
		if (unlikely(!putq(q, mp))) {
			swerr();
			freemsg(mp);
		}
		return (0);
	}
	if (unlikely(db_type >= QPCTL)) {
		if (likely(db_type != M_FLUSH)) {
			putnext(q, mp);
			return (0);
		} else {
			if (mp->b_rptr[0] & FLUSHR) {
				if (mp->b_rptr[0] & FLUSHBAND)
					flushband(q, mp->b_rptr[1], FLUSHDATA);
				else
					flushq(q, FLUSHDATA);
				mp->b_rptr[0] &= ~FLUSHR;
			}
			putnext(q, mp);
			return (0);
		}
	} else if (unlikely(db_type == M_IOCTL)) {
		union ioctypes *ioc = (typeof(ioc)) mp->b_rptr;
		int err = EINVAL;
		int rval = 0;
		int count = 0;
		int cmd;
		uint32_t *args = (typeof(args)) (mp->b_cont ? mp->b_cont->b_rptr : NULL);

		count = ioc->iocblk.ioc_count;

		switch ((cmd = ioc->iocblk.ioc_cmd)) {
		case 0:	/* none */
			priv->mode = cmd;
			__printd(("%s: set mode to NONE\n", MOD_NAME));
			goto ack;
		case 1:	/* discard */
			priv->mode = cmd;
			__printd(("%s: set mode to DISCARD\n", MOD_NAME));
			goto ack;
		case 2:	/* echo */
			priv->mode = cmd;
			__printd(("%s: set mode to ECHO\n", MOD_NAME));
			goto ack;
		case 3:	/* generator 1 stream */
			if (count == TRANSPARENT)
				goto nak;
			if (args == NULL || count < 2 * sizeof(uint32_t))
				goto nak;
			priv->mode = cmd;
			priv->number = args[0];
			priv->burst = args[1];
			priv->streams = 1;
			__printd(("%s: set mode to GENERATE, number = %u, burst = %u, streams = %u\n", MOD_NAME, priv->number, priv->burst, priv->streams));
			count = 0;
			goto ack;
		case 4:	/* generator N streams */
			if (count == TRANSPARENT)
				goto nak;
			if (args == NULL || count < 3 * sizeof(uint32_t))
				goto nak;
			priv->mode = cmd;
			priv->number = args[0];
			priv->burst = args[1];
			priv->streams = args[2];
			__printd(("%s: set mode to GENERATE, number = %u, burst = %u, streams = %u\n", MOD_NAME, priv->number, priv->burst, priv->streams));
			count = 0;
			goto ack;
		case 5:	/* collect stats with reset */
			if (count == TRANSPARENT)
				goto nak;
			if (args == NULL) {
				if ((mp->b_cont = allocb(2 * sizeof(uint32_t), BPRI_MED)) == NULL)
					goto nak;
				args = (typeof(args)) (mp->b_cont ? mp->b_cont->b_rptr : NULL);
			}
			args[0] = xchg(&priv->count, 0);
			args[1] = xchg(&priv->bytes, 0);
			count = 2 * sizeof(uint32_t);
			mp->b_cont->b_wptr = mp->b_cont->b_rptr + count;
			goto ack;
		}
		goto nak;
	      ack:
		mp->b_datap->db_type = M_IOCACK;
		ioc->iocblk.ioc_count = count;
		ioc->iocblk.ioc_rval = rval;
		ioc->iocblk.ioc_error = 0;
		qreply(q, mp);
		return (0);
	      nak:
		mp->b_datap->db_type = M_IOCNAK;
		ioc->iocblk.ioc_count = count;
		ioc->iocblk.ioc_rval = -1;
		ioc->iocblk.ioc_error = err;
		qreply(q, mp);
		return (0);
	} else if (db_type == M_PROTO) {
		switch (priv->mode) {
		case 0:
		case 1:
		case 2:
			break;
		case 3:	/* generator 1 stream */
		case 4:	/* generator multiple streams */
			if (likely(mp->b_wptr >= mp->b_rptr + sizeof(t_scalar_t))) {
				switch (*((t_scalar_t *) mp->b_rptr)) {
				case T_OPTDATA_REQ:
				case T_DATA_REQ:
				case T_EXDATA_REQ:
				{
					/* save message for later */
					if (priv->mp != NULL)
						freemsg(xchg(&priv->mp, mp));
					priv->mp = mp;
					qenable(q);
					return (0);
				}
				}
			}
			break;
		}
	}
	if (likely(bcanputnext(q, mp->b_band))) {
		putnext(q, mp);
	} else if (unlikely(!putq(q, mp))) {
		swerr();
		freemsg(mp);
	}
	return (0);
}

static streamscall int
tpiperf_wsrv(queue_t *q)
{
	mblk_t *mp;

	while (likely((mp = getq(q)) != NULL)) {
		/* the only messages that are placed on the queue are messages that are to be
		   passed downstream without modification. */
		if (likely(bcanputnext(q, mp->b_band) || mp->b_datap->db_type >= QPCTL)) {
			putnext(q, mp);
			continue;
		}
		if (unlikely(!putbq(q, mp))) {
			swerr();
			freemsg(mp);
		}
		break;
	}
	{
		struct tpiperf *priv = (typeof(priv)) q->q_ptr;

		switch (priv->mode) {
		case 0:	/* none */
		case 1:	/* discard */
		case 2:	/* echo */
			break;
		case 3:	/* generator */
			if ((mp = priv->mp)) {
				mblk_t *dp;
				uint burst = 0;

				/* start generating immediately */
				for (;;) {
					if (likely((dp = dupmsg(mp)) != NULL)) {
						if (likely(mp->b_cont != NULL))
							priv->bytes += msgsize(mp->b_cont);
						priv->count++;
						burst++;
						if (likely(bcanputnext(q, dp->b_band))) {
							putnext(q, dp);
							if (priv->count > priv->number) {
								priv->mode = 0;
								if ((dp =
								     allocb(2 * sizeof(t_scalar_t),
									    BPRI_MED))) {
									dp->b_datap->db_type =
									    M_PROTO;
									*((t_scalar_t *) dp->
									  b_wptr) = T_DISCON_REQ;
									dp->b_wptr +=
									    sizeof(t_scalar_t);
									*((t_scalar_t *) dp->
									  b_wptr) = 0;
									dp->b_wptr +=
									    sizeof(t_scalar_t);
									putnext(q, dp);
								}
								if ((dp =
								     allocb(3 * sizeof(t_scalar_t),
									    BPRI_MED))) {
									dp->b_datap->db_type =
									    M_PROTO;
									*((t_scalar_t *) dp->
									  b_wptr) = T_DISCON_IND;
									dp->b_wptr +=
									    sizeof(t_scalar_t);
									*((t_scalar_t *) dp->
									  b_wptr) = 0;
									dp->b_wptr +=
									    sizeof(t_scalar_t);
									*((t_scalar_t *) dp->
									  b_wptr) = 0;
									dp->b_wptr +=
									    sizeof(t_scalar_t);
									put(priv->rq, dp);
								}
								break;
							} else if (burst > priv->burst) {
								qenable(q);
								break;
							}
							continue;
						}
						if (unlikely(!putbq(q, dp))) {
							swerr();
							freemsg(dp);
						}
						break;
					}
				}
				break;
			}
			break;
		case 4:	/* generator multiple streams */
			/* FIXME: this is broken: mp is NULL! */
			if (likely(mp->b_wptr >= mp->b_rptr + sizeof(t_scalar_t))) {
				switch (*((t_scalar_t *) mp->b_rptr)) {
				case T_OPTDATA_IND:
				case T_DATA_IND:
				case T_EXDATA_IND:
				{
					mblk_t *bp, *dp;

					/* save message for later */
					if (priv->mp != NULL)
						freemsg(xchg(&priv->mp, mp));
					/* start generating immediately */
					for (;;) {
						if (likely((bp = copyb(mp)) != NULL)) {
							if (likely
							    ((dp = mp->b_cont) == NULL
							     || (dp = dupmsg(dp)) != NULL)) {
								if (likely
								    ((bp->b_cont = dp) != NULL))
									priv->bytes += msgsize(dp);
								priv->count++;
								if (likely(bcanput(q, bp->b_band))) {
									put(q, bp);
									continue;
								}
							} else {
								freeb(bp);
								if (mp->b_cont) {
									/* need bufcall */
									if (priv->bcid)
										unbufcall(xchg
											  (&priv->
											   bcid,
											   0));
									priv->bcid =
									    bufcall(msgsize
										    (mp->b_cont),
										    BPRI_MED,
										    &tpiperf_bufcall,
										    (long) q);
								}
							}
						} else {
							/* need bufcall */
							if (priv->bcid)
								unbufcall(xchg(&priv->bcid, 0));
							priv->bcid =
							    bufcall(mp->b_wptr - mp->b_rptr,
								    BPRI_MED, &tpiperf_bufcall,
								    (long) q);
						}
						if (unlikely(!putbq(q, bp))) {
							swerr();
							freemsg(bp);
						}
						break;
					}
					return (0);
				}
				}
			}
			break;
		}
	}
	return (0);
}

static streamscall int
tpiperf_rput(queue_t *q, mblk_t *mp)
{
	struct tpiperf *priv = (typeof(priv)) q->q_ptr;
	int db_type = mp->b_datap->db_type;

	if (unlikely((q->q_first || (q->q_flag & QSVCBUSY)) && db_type < QPCTL)) {
		if (unlikely(!putq(q, mp))) {
			swerr();
			freemsg(mp);
		}
		return (0);
	}
	if (likely(db_type == M_PROTO)) {
		switch (priv->mode) {
		case 0:	/* nothing */
			break;
		case 1:	/* discard */
		case 3:	/* generator */
			if (likely(mp->b_wptr >= mp->b_rptr + sizeof(t_scalar_t))) {
				switch (*((t_scalar_t *) mp->b_rptr)) {
				case T_OPTDATA_IND:
				case T_DATA_IND:
				case T_EXDATA_IND:
					freemsg(mp);
					return (0);
				}
			}
			break;
		case 2:	/* echo */
			if (likely(mp->b_wptr >= mp->b_rptr + sizeof(t_scalar_t))) {
				switch (*((t_scalar_t *) mp->b_rptr)) {
				case T_OPTDATA_IND:
					*((t_scalar_t *) mp->b_rptr) = T_OPTDATA_REQ;
					goto echo_it;
				case T_DATA_IND:
					*((t_scalar_t *) mp->b_rptr) = T_DATA_REQ;
					goto echo_it;
				case T_EXDATA_IND:
					*((t_scalar_t *) mp->b_rptr) = T_EXDATA_REQ;
					goto echo_it;
				}
			}
			break;
		}
	} else if (unlikely(db_type >= QPCTL)) {
		if (likely(db_type != M_FLUSH)) {
			putnext(q, mp);
			return (0);
		} else {
			if (mp->b_rptr[0] & FLUSHR) {
				if (mp->b_rptr[0] & FLUSHBAND)
					flushband(q, mp->b_rptr[1], FLUSHDATA);
				else
					flushq(q, FLUSHDATA);
				mp->b_rptr[0] &= ~FLUSHR;
			}
			putnext(q, mp);
			return (0);
		}
	}
	goto pass_it;
      echo_it:
	if (likely(bcanputnext(priv->wq, mp->b_band))) {
		putnext(priv->wq, mp);
	} else if (unlikely(!putq(priv->wq, mp))) {
		swerr();
		freemsg(mp);
	}
	return (0);
      pass_it:
	if (likely(bcanputnext(q, mp->b_band))) {
		putnext(q, mp);
	} else if (unlikely(!putq(q, mp))) {
		swerr();
		freemsg(mp);
	}
	return (0);
}

static streamscall int
tpiperf_rsrv(queue_t *q)
{
	mblk_t *mp;

	while (likely((mp = getq(q)) != NULL)) {
		/* the only messages that are placed on the queue are messages that are to be
		   passed upstream without modification. */
		if (likely(bcanputnext(q, mp->b_band) || mp->b_datap->db_type >= QPCTL)) {
			putnext(q, mp);
			continue;
		} else if (unlikely(!putbq(q, mp))) {
			swerr();
			freemsg(mp);
			break;
		}
	}
	return (0);
}

/*
 *  =========================================================================
 *
 *  OPEN and CLOSE
 *
 *  =========================================================================
 *  Note that LiS does not do module counting, therefore we have to do it.
 */

static streamscall int
tpiperf_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *crp)
{
	int err = 0;

	if (q->q_ptr != NULL)
		goto quit;	/* already open */
	err = ENXIO;
	if (sflag != MODOPEN || WR(q)->q_next == NULL)
		goto quit;
	err = ENOMEM;
	if (!tpiperf_alloc_priv(q))
		goto quit;
	err = 0;
	qprocson(q);
      quit:
	return (err);
}

static streamscall int
tpiperf_close(queue_t *q, int oflag, cred_t *crp)
{
	qprocsoff(q);
	tpiperf_free_priv(q);
	return (0);
}

/*
 *  =========================================================================
 *
 *  Registration and initialization
 *
 *  =========================================================================
 */
#ifdef LINUX
/*
 *  Linux Registration
 *  -------------------------------------------------------------------------
 */

unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for the TIMOD module. (0 for allocation.)");

/*
 *  Linux Fast-STREAMS Registration
 *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 */
STATIC struct fmodsw tpiperf_fmod = {
	.f_name = MOD_NAME,
	.f_str = &tpiperfinfo,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

STATIC int
tpiperf_register_strmod(void)
{
	int err;

	if ((err = register_strmod(&tpiperf_fmod)) < 0)
		return (err);
	return (0);
}

STATIC int
tpiperf_unregister_strmod(void)
{
	int err;

	if ((err = unregister_strmod(&tpiperf_fmod)) < 0)
		return (err);
	return (0);
}

MODULE_STATIC int __init
tpiperfinit(void)
{
	int err;

	cmn_err(CE_NOTE, MOD_BANNER);	/* banner message */
	if ((err = tpiperf_init_caches())) {
		cmn_err(CE_WARN, "%s: could not init caches, err = %d", MOD_NAME, err);
		return (err);
	}
	if ((err = tpiperf_register_strmod())) {
		cmn_err(CE_WARN, "%s: could not register module, err = %d", MOD_NAME, err);
		tpiperf_term_caches();
		return (err);
	}
	if (modid == 0)
		modid = err;
	return (0);
}

MODULE_STATIC void __exit
tpiperfterminate(void)
{
	int err;

	if ((err = tpiperf_unregister_strmod()))
		cmn_err(CE_WARN, "%s: could not unregister module", MOD_NAME);
	if ((err = tpiperf_term_caches()))
		cmn_err(CE_WARN, "%s: could not terminate caches", MOD_NAME);
	return;
}

/*
 *  Linux Kernel Module Initialization
 *  -------------------------------------------------------------------------
 */
module_init(tpiperfinit);
module_exit(tpiperfterminate);

#endif				/* LINUX */
